executors/keptn/pkg/keptn/keptn.go

```diff
@@ -0,0 +1,329 @@
+	jobExecutorFilename = "config.yaml"
+	jobExecutorURI      = "job/config.yaml"
+
+	ShipyardFileName    string = "shipyard.yaml"
+	KeptnConfigFileName string = "keptn-config.json"
+)
+
+var (
+	ErrFailedDeleteProject = fmt.Errorf("failed to delete project")
+	ErrEvaluationFailed    = fmt.Errorf("evaluation result shows failure")
+)
+
+func resourceNameToURI(fname string) string {
+	switch fname {
+	case sliFilename:
+		return sliURI
+	case sloFilename:
+		return sloURI
+	case jobExecutorFilename:
+		return jobExecutorURI
+	default:
+		return fname
+	}
+}
+
+type Git struct {
+	URL   string
+	Token string
+	User  string
+}
+
+type Keptn struct {
+	url             string
+	git             *Git
+	token           *string
+	apiHandler      *apiutils.APIHandler
+	resourceHandler *apiutils.ResourceHandler
+	projectHandler  *apiutils.ProjectHandler
+	eventHandler    *apiutils.EventHandler
+}
+
+func New(url, namespace, secretName string, git *Git) (*Keptn, error) {
+	if git == nil {
+		log.Printf("No upstream git server provided. Using in-built git server")
+	}
+
+	// get token from secret
+	t, err := keptnk8sutils.GetKeptnAPITokenFromSecret(true, namespace, secretName)
+	if err != nil {
+		return nil, err
+	}
+
+	// authenticate with the api server
+	auth := apiutils.NewAuthenticatedAuthHandler(url, t, "x-token", nil, "http")
+	if _, kErr := auth.Authenticate(); kErr != nil {
+		err = fmt.Errorf("failed to authenticate with err: %v", kErr)
+		log.Printf("failed to authenticate with err : %v", kErr.GetMessage())
+		return nil, err
+	}
+
+	apiHandler := apiutils.NewAuthenticatedAPIHandler(url, t, "x-token", nil, "http")
+	resourceHandler := apiutils.NewAuthenticatedResourceHandler(url, t, "x-token", nil, "http")
+	projectHandler := apiutils.NewAuthenticatedProjectHandler(url, t, "x-token", nil, "http")
+	eventHandler := apiutils.NewAuthenticatedEventHandler(url, t, "x-token", nil, "http")
+
+	return &Keptn{
+		url:             url,
+		token:           &t,
+		apiHandler:      apiHandler,
+		resourceHandler: resourceHandler,
+		projectHandler:  projectHandler,
+		eventHandler:    eventHandler,
+		git:             git,
+	}, nil
+}
+
+func (k *Keptn) CreateOrUpdateProject(project string, shipyard string) error {
+	encodedShipyardContent := base64.StdEncoding.EncodeToString([]byte(shipyard))
+	projectInfo := apimodels.CreateProject{
+		Name:     &project,
+		Shipyard: &encodedShipyardContent,
+	}
+
+	projectGetInfo := apimodels.Project{
+		ProjectName: project,
+	}
+
+	if k.git != nil {
+		projectInfo.GitRemoteURL = k.git.URL
+		projectInfo.GitToken = k.git.Token
+		projectInfo.GitUser = k.git.User
+
+		projectGetInfo.GitRemoteURI = k.git.URL
+		projectGetInfo.GitToken = k.git.Token
+		projectGetInfo.GitUser = k.git.User
+	}
+
+	if _, kErr := k.projectHandler.GetProject(apimodels.Project{
+		ProjectName:     project,
+		ShipyardVersion: shipyard,
+		Stages:          []*apimodels.Stage{},
+	}); kErr == nil {
```

Comment:
What happens if the `Get` call fails due to an actual error but the job exists?

Reply:
Good point - will update. Need to check if keptn has specific error codes/messages that state what the error type was.